File: apps/visualizer/backend/api/descriptions.py
import json
import os

from flask import Blueprint, jsonify, request
from lightroom_tagger.core.database import (
    get_all_images_with_descriptions,
    get_image_description,
)
from lightroom_tagger.core.description_service import (
    describe_instagram_image,
    describe_matched_image,
)
from utils.db import with_db
from utils.responses import error_server_error

bp = Blueprint('descriptions', __name__)

_JSON_COLS = ('composition', 'perspectives', 'technical', 'subjects')


@bp.route('/', methods=['GET'])
@with_db
def list_descriptions(db):
    """List images with their AI descriptions."""
    try:
        image_type = request.args.get('image_type')  # catalog, instagram, or None for both
        described_only = request.args.get('described_only', 'false') == 'true'
        limit = request.args.get('limit', 50, type=int)
        offset = request.args.get('offset', 0, type=int)

        items, total = get_all_images_with_descriptions(
            db, image_type=image_type, described_only=described_only,
            limit=limit, offset=offset,
        )

        return jsonify({
            'total': total,
            'items': items,
            'pagination': {
                'offset': offset,
                'limit': limit,
                'current_page': (offset // limit) + 1,
                'total_pages': (total + limit - 1) // limit if total else 0,
                'has_more': (offset + limit) < total,
            },
        })
    except Exception as e:
        return error_server_error(str(e))


@bp.route('/<path:image_key>', methods=['GET'])
@with_db
def get_description(db, image_key):
    """Get full description for a single image."""
    try:
        desc = get_image_description(db, image_key)
        if not desc:
            return jsonify({'description': None})
        return jsonify({'description': _deserialize(desc)})
    except Exception as e:
        return error_server_error(str(e))


@bp.route('/<path:image_key>/generate', methods=['POST'])
@with_db
def generate_description(db, image_key):
    """Generate AI description for a single image."""
    from lightroom_tagger.core.provider_errors import (
        AuthenticationError,
        ConnectionError,
        ModelUnavailableError,
        RateLimitError,
    )

    old_model_env = os.environ.get('DESCRIPTION_VISION_MODEL')
    provider_id = None
    try:
        data = request.json or {}
        image_type = data.get('image_type', 'catalog')
        force = data.get('force', False)
        model = data.get('model')
        provider_id = data.get('provider_id')
        provider_model = data.get('provider_model')

        if model and not provider_id:
            os.environ['DESCRIPTION_VISION_MODEL'] = model

        try:
            if image_type == 'catalog':
                generated = describe_matched_image(
                    db, image_key, force=force,
                    provider_id=provider_id, model=provider_model,
                )
            elif image_type == 'instagram':
                generated = describe_instagram_image(
                    db, image_key, force=force,
                    provider_id=provider_id, model=provider_model,
                )
            else:
                return jsonify({'error': f'Invalid image_type: {image_type}'}), 400
        except KeyError:
            if not provider_id:
                raise
            return jsonify({
                'error': 'invalid_provider',
                'message': f'Unknown provider: {provider_id}',
            }), 400

        if not generated:
            existing = get_image_description(db, image_key)
            if existing:
                return jsonify({'generated': False, 'description': _deserialize(existing)})
            return jsonify({'generated': False, 'description': None})

        desc = get_image_description(db, image_key)
        return jsonify({'generated': True, 'description': _deserialize(desc) if desc else None})
    except RateLimitError as e:
        return jsonify({
            'error': 'rate_limit',
            'message': str(e),
            'provider': getattr(e, 'provider', None) or provider_id,
        }), 429
    except AuthenticationError as e:
        return jsonify({
            'error': 'auth_error',
            'message': str(e),
            'provider': getattr(e, 'provider', None) or provider_id,
        }), 401
    except (ModelUnavailableError, ConnectionError) as e:
        return jsonify({
            'error': 'provider_unavailable',
            'message': str(e),
            'provider': getattr(e, 'provider', None) or provider_id,
        }), 503
    except Exception as e:
        return error_server_error(str(e))
    finally:
        if old_model_env is not None:
            os.environ['DESCRIPTION_VISION_MODEL'] = old_model_env
        else:
            os.environ.pop('DESCRIPTION_VISION_MODEL', None)


def _deserialize(row: dict) -> dict:
    out = dict(row)
    for col in _JSON_COLS:
        val = out.get(col)
        if isinstance(val, str):
            try:
                out[col] = json.loads(val)
            except (json.JSONDecodeError, TypeError):
                pass
    return out

File: apps/visualizer/backend/src/api/descriptions.ts
/**
 * Image description routes.
 */
import { createRoute, z } from '@hono/zod-openapi';
import {
  getAllImagesWithDescriptions,
  getImageDescription,
} from '../db/library/descriptions.js';
import { libraryDb, type LibraryEnv } from './library-db.js';
import {
  AuthenticationError,
  ModelUnavailableError,
  ProviderConnectionError,
  RateLimitError,
} from '../providers/errors.js';
import { UnknownProviderError } from '../providers/registry.js';
import { HttpError } from '../utils/responses.js';
import { describeMatchedImage } from '../vision/description-service.js';
import { createOpenApiApp } from './openapi.js';
import { jsonBody, redirectToTrailingSlash, withValidationError } from './route-helpers.js';
import { ErrorBody } from './schemas/errors.js';
import {
  DescriptionGenerateRequest,
  DescriptionGenerateResponse,
  DescriptionGetResponse,
  DescriptionProviderError,
  DescriptionsListResponse,
  ImageDescription,
} from './schemas/descriptions.js';

export const descriptionsRoutes = createOpenApiApp<LibraryEnv>();

// Only `POST /{image_key}/generate` writes; the two list/detail reads stay read-only.
descriptionsRoutes.use('/descriptions/*', libraryDb({ writeForMethods: ['POST'] }));
redirectToTrailingSlash(descriptionsRoutes, '/descriptions');

/**
 * `limit`/`offset` use loose int parsing: an unparseable value falls back to the
 * default rather than erroring. Unlike the shared pagination helper, no bounds are
 * imposed here.
 */
function intArg(raw: string | undefined, fallback: number): number {
  if (raw === undefined || raw === '') return fallback;
  const n = Number(raw);
  return Number.isInteger(n) ? n : fallback;
}

const listRoute = createRoute({
  method: 'get',
  path: '/descriptions/',
  tags: ['descriptions'],
  responses: withValidationError({
    200: { description: 'Images with descriptions', content: jsonBody(DescriptionsListResponse) },
  }),
});

descriptionsRoutes.openapi(listRoute, (c) => {
  const imageType = c.req.query('image_type');
  // Catalog-only since #218; anything else is a client error, not empty results.
  // Thrown rather than returned: the OpenAPI contract documents only 200 and 422.
  // See `HttpError`.
  if (imageType !== undefined && imageType !== '' && imageType !== 'catalog') {
    throw new HttpError(400, `Invalid image_type: ${imageType}`);
  }

  const describedOnly = c.req.query('described_only') === 'true';
  const limit = intArg(c.req.query('limit'), 50);
  const offset = intArg(c.req.query('offset'), 0);

  const { items, total } = getAllImagesWithDescriptions(c.get('libraryDb'), {
    describedOnly,
    limit,
    offset,
  });

  return c.json(
    {
      total,
      items,
      pagination: {
        offset,
        limit,
        current_page: Math.floor(offset / limit) + 1,
        // Empty result yields 0 pages, not 1.
        total_pages: total ? Math.floor((total + limit - 1) / limit) : 0,
        has_more: offset + limit < total,
      },
    },
    200,
  );
});

const detailRoute = createRoute({
  method: 'get',
  path: '/descriptions/{image_key}',
  tags: ['descriptions'],
  request: { params: z.object({ image_key: z.string() }) },
  responses: withValidationError({
    200: { description: 'One description, or null', content: jsonBody(DescriptionGetResponse) },
  }),
});

descriptionsRoutes.openapi(detailRoute, (c) => {
  // A missing description is `{"description": null}` with a 200, not a 404 —
  // the UI renders an empty panel rather than treating it as an error.
  const desc = getImageDescription(c.get('libraryDb'), c.req.param('image_key'));
  return c.json({ description: desc as z.infer<typeof DescriptionGetResponse>['description'] }, 200);
});

type GenerateDescription = z.infer<typeof ImageDescription> | null;

const generateRoute = createRoute({
  method: 'post',
  path: '/descriptions/{image_key}/generate',
  tags: ['descriptions'],
  summary: 'Generate AI description for a single image.',
  request: {
    params: z.object({ image_key: z.string() }),
    body: { required: true, content: jsonBody(DescriptionGenerateRequest) },
  },
  responses: withValidationError({
    200: { description: 'OK', content: jsonBody(DescriptionGenerateResponse) },
    400: { description: 'Bad Request', content: jsonBody(ErrorBody) },
    401: { description: 'Unauthorized', content: jsonBody(DescriptionProviderError) },
    429: { description: 'Too Many Requests', content: jsonBody(DescriptionProviderError) },
    503: { description: 'Service Unavailable', content: jsonBody(DescriptionProviderError) },
  }),
});

descriptionsRoutes.openapi(generateRoute, async (c) => {
  const db = c.get('libraryDb');
  const imageKey = c.req.param('image_key');
  const body = c.req.valid('json');
  const providerId = body.provider_id ?? null;

  // Catalog-only since #218. Validated here rather than as a Zod enum so bad values
  // return 400, not 422.
  if (body.image_type !== '' && body.image_type !== 'catalog') {
    return c.json({ error: `Invalid image_type: ${body.image_type}` }, 400);
  }

  // The model is passed as an argument instead of via env var to avoid races between
  // concurrent requests. `provider_model` wins over `model`; bare `model` is ignored
  // when an explicit `provider_id` is given.
  const model = body.provider_model ?? (providerId ? null : body.model ?? null);

  const respond = (generated: boolean, desc: unknown) =>
    c.json({ generated, description: (desc ?? null) as GenerateDescription }, 200);

  try {
    const outcome = await describeMatchedImage(db, imageKey, {
      force: body.force,
      providerId,
      model,
    });

    // A skipped or failed outcome still returns whatever description already
    // exists, so the UI can show the old text instead of blanking the panel.
    if (!outcome.wrote) return respond(false, getImageDescription(db, imageKey));
    return respond(true, getImageDescription(db, imageKey));
  } catch (e) {
    // Without `provider_id`, an unknown provider from config is a server fault (500).
    if (e instanceof UnknownProviderError) {
      if (!providerId) throw e;
      return c.json(
        { error: 'invalid_provider', message: `Unknown provider: ${providerId}` },
        400,
      );
    }
    // `provider` falls back to the requested id: the exception often lacks it
    // when the failure happened before any provider was contacted, and the UI
    // needs to name the provider it asked for.
    if (e instanceof RateLimitError) {
      return c.json(
        { error: 'rate_limit', message: e.message, provider: e.provider ?? providerId },
        429,
      );
    }
    if (e instanceof AuthenticationError) {
      return c.json(
        { error: 'auth_error', message: e.message, provider: e.provider ?? providerId },
        401,
      );
    }
    if (e instanceof ModelUnavailableError || e instanceof ProviderConnectionError) {
      return c.json(
        { error: 'provider_unavailable', message: e.message, provider: e.provider ?? providerId },
        503,
      );
    }
    // Unhandled provider errors become 500 via `app.onError`.
    throw e;
  }
});
